### PR TITLE
Add OneHotEncoder Ability to collapse infrequent values issue 26

### PR DESCRIPTION
**Pull request by adithyabsk**
### Related Issue
Fixes issue #26 

### Description
* Add UncommonRemover internal transformer to remove infrequent values in a categorical column
* Re-format smart transformers to have constructors and document parameters
* Add UncommonRemover as a preprocessing step to OneHotEncoder
* Add relevant tests
* Add field to check_df to validate single column DataFrame and add
  relevant tests
* Update smart encoder to make a recursive call to itself to make sure
  that UncommonRemover is evaluated before an encoding method is picked

### CODE REVIEW

File: foreshadow/transformers/internals/uncommonremover.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        self.replacement = replacement
+
+    def fit(self, X, y=None):
+        """
+        Finds the uncommon values and sets the replacement value
+        """
+        X = check_array(X, dtype=None)
+        self.values_, counts = np.unique(X, return_counts=True)
+        self.merge_values_ = self.values_[np.where(counts <= self.threshold * X.size)]
+        if self.replacement is None:
```
Comment (jichaogp) [inline foreshadow/transformers/internals/uncommonremover.py]: I feel the logic to set the default value for replacement here could be problematic. If it is a string column, we could end up with a very long name at the end. If it is numeric, we won't have a way to tell a category is the result of merging. I suggest we simply set the default value to 'DummyEncoder_Other'.

File: foreshadow/transformers/smart.py [hunk 2]
```diff
@@ -53,24 +60,39 @@ class Encoder(SmartTransformer):
...
         unique_count = len(data.value_counts())
         if self.y_var:
             return LabelEncoder()
-        if unique_count <= unique_num_cutoff:
-            return OneHotEncoder(
-                cols=[col_name],
-                return_df=True,
-                use_cat_names=True,
-                handle_unknown="ignore",
+        if unique_count <= self.unique_num_cutoff:
```
Comment (jichaogp) [inline foreshadow/transformers/smart.py]: What if after the number of categories is less than 30 after merging UncommonRemover?

File: foreshadow/transformers/internals/uncommonremover.py (hunk 1 above)
Comment (adithyabsk) [reply foreshadow/transformers/internals/uncommonremover.py]: I wanted to leave as much information in the data as possible but if you think it is best to go that route, I can do that! I'll make the change.

File: foreshadow/transformers/smart.py (hunk 2 above)
Comment (adithyabsk) [reply foreshadow/transformers/smart.py]: Yep I can make it so that it will resolve after uncommon processes

**Comment by adithyabsk** [comment]
I'm perfectly happy to change the way it works. What would you think the ideal implementation would be. Another implementation I considered was a partial fit of UncommonRemover and then based on the output a selection of a specific pipeline. I'd don't think there should be any problems with serialization but @robocoder99 can confirm that.

**Comment by jichaogp** [comment]
Hi @adithyabsk , please push the change. We can decide whether we want to address my comments after we apply the framework to some open dataset based on our observation.

**Comment by adithyabsk** [comment]
> Hi @adithyabsk , please push the change. We can decide whether we want to address my comments after we apply the framework to some open dataset based on our observation.

I ended up having to implement a solution similar to your suggestion in order to address a merge conflict that arose from @robocoder99's addition to the smart encoder.

**Pull request merged**